File: sentimental-credit/credit.py
import re


def main():
    # get card num from user
    cc_number = get_card()
    dig_length = len(cc_number)

    # Run card number through checks
    amex = isValidAmex(cc_number, dig_length)
    master = isValidMaster(cc_number, dig_length)
    visa = isValidVisa(cc_number, dig_length)
    
    # Run card though secret formula
    sum_every_other_digit = sum_of_alternate_digits(cc_number)

    # Check the last digit of the card and return card type if applicable
    if sum_every_other_digit % 10 != 0:
        print("INVALID")
    elif amex == True:
        print("AMEX")
    elif master == True:
        print("MASTERCARD")
    elif visa == True:
        print("VISA")
    else:
        print("INVALID")


# Check if card number is an Amex card
def isValidAmex(num, len):
    first_two_dig = num[0] + num[1]
    if len == 15 and (first_two_dig == '34' or first_two_dig == '37'):
        return True
    else:
        return False


# Check if card number is a Master card
def isValidMaster(num, len):
    first_two_dig = num[0] + num[1]
    if len == 16 and 50 <= int(first_two_dig) < 56:
        return True
    else:
        return False


# Check if card number is a Visa card
def isValidVisa(num, len):
    first_dig = num[0]
    if (len == 16 or len == 13) and first_dig == '4':
        return True
    else:
        return False


# Formula for getting the sum of every other digit from the 2nd to last number of the credit card
def sum_of_alternate_digits(num):
    total = 0
    for i in range(len(num) - 2, -1, -2):
        digit = int(num[i]) * 2
        if digit > 9:
            digit = sum(int(d) for d in str(digit))
        total += digit
    for i in range(len(num) - 1, -1, -2):
        total += int(num[i])
    return total


# Prompt user for card info
def get_card():
    while True:
        card = int(input("What is your Credit Card number? "))
        if card > 0:
            return str(card)


main()

File: credit/credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <string.h>

// Converts card number to a string so we can find the length
int length(long card_num);

// Formula for getting the sum of every other digit from the 2nd to last number of the credit card
int sum_of_alternate_digits(long card_num);

// Doubles the last digit and adds the digits together (not the integer)
int doubled_and_sum(long last_digit);

// Checks if the card number has the qualities of an amex card
bool isValidAmex(long card_num, int num_length);

// Checks if the card number has the qualities of a Master card
bool isValidMaster(long card_num, int num_length);

// Checks if the card number has the qualities of a visa card
bool isValidVisa(long card_num, int num_length);

int main(void)
{
    long cc_number;
    // Ask user for CC number
    do
    {
        cc_number = get_long("What is your Credit Card number? ");
    }
    while (cc_number < 0);

    // Assigning variables for functions
    int dig_length = length(cc_number);
    bool amex = isValidAmex(cc_number, dig_length);
    bool master = isValidMaster(cc_number, dig_length);
    bool visa = isValidVisa(cc_number, dig_length);
    int sum_every_other_digit = sum_of_alternate_digits(cc_number);

    // Runs the CC number through several functions to verify type and validity
    if (sum_every_other_digit % 10 != 0)
    {
        printf("INVALID\n");
    }
    else if (amex == true)
    {
        printf("AMEX\n");
    }
    else if (master == true)
    {
        printf("MASTERCARD\n");
    }
    else if (visa == true)
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
    }

    // print tests to make sure numbers were valid
    // printf("%i", sum_every_other_digit);
    // printf("\n");
    // printf("%d", length(cc_number));
    // printf("\n");
}









// Formula for getting the sum of every other digit from the 2nd to last number of the credit card
int sum_of_alternate_digits(long card_num)
{
    int sum = 0;
    bool isOtherDigit = false;
    while (card_num > 0)
    {
        if (isOtherDigit == true)
        {
            int final_digit = card_num % 10;
            int product = doubled_and_sum(final_digit);
            sum = sum + product;
        }
        else
        {
            int final_digit = card_num % 10;
            sum = sum + final_digit;
        }
        isOtherDigit = !isOtherDigit;
        card_num = card_num / 10;
    }
    return sum;

}

// Doubles the last digit and adds the digits together (not the integer)
int doubled_and_sum(long last_digit)
{
    int sum = 0;
    int doubled = last_digit * 2;
    while (doubled > 0)
    {
        int last_digit_after_doubled = doubled % 10;
        sum = sum + last_digit_after_doubled;
        doubled = doubled / 10;
    }
    return sum;

}

// Converts card number to a string so we can find the length
int length(long card_num)
{
    char str[20];
    int len;
    len = snprintf(str, 20, "%ld", card_num);
    len = strlen(str);
    return len;
}

// Checks if the card number has the qualities of an amex card
bool isValidAmex(long card_num, int num_length)
{
    int first_two_dig = (card_num / 10000000000000);
    if ((num_length == 15) && (first_two_dig == 34 || first_two_dig == 37))
    {
        return true;
    }
    else
    {
        return false;
    }
}

// Checks if the card number has the qualities of a Master card
bool isValidMaster(long card_num, int num_length)
{
    int first_two_dig = (card_num / 100000000000000);
    if ((num_length == 16) && (first_two_dig > 50 && first_two_dig < 56))
    {
        return true;
    }
    else
    {
        return false;
    }
}

// Checks if the card number has the qualities of a visa card
bool isValidVisa(long card_num, int num_length)
{
    int first_dig_a = (card_num / 1000000000000000);
    int first_dig_b = (card_num / 1000000000000);
    if ((num_length == 13 || num_length == 16) && (first_dig_a == 4 || first_dig_b == 4))
    {
        return true;
    }
    else
    {
        return false;
    }
}